src/rgw/rgw_sal_motr.cc

```diff
@@ -905,25 +905,50 @@ std::unique_ptr<Object> MotrBucket::get_object(const rgw_obj_key& k)
 int MotrBucket::list(const DoutPrefixProvider *dpp, ListParams& params, int max, ListResults& results, optional_yield y)
 {
   int rc;
+  if (max <= 0)  // Return an emtpy response.
+    return 0;
+  max++;  // Fetch an extra key if available to ensure presence of next obj.
   vector<string> keys(max);
   vector<bufferlist> vals(max);
 
   ldpp_dout(dpp, 20) << "bucket=" << info.bucket.name
                     << " prefix=" << params.prefix
                     << " marker=" << params.marker
-                    << " max=" << max << dendl;
+                    << " max=" << max-1 << dendl;
 
   // Retrieve all `max` number of pairs.
   string bucket_index_iname = "motr.rgw.bucket.index." + info.bucket.name;
-  keys[0] = params.marker.empty() ? params.prefix :
-                                    params.marker.to_str();
+
+  // Modify the marker based on its type
+  keys[0] = params.prefix;
+  if (!params.marker.empty()) {
+    keys[0] = params.marker.to_str();
+    // Get the position of delimiter string
+    int delim_pos = keys[0].find(params.delim, params.prefix.length());
+    // If delimiter is present at the very end, append "\xff" to skip all
+    // the dir entries, else append " " to skip the maker key.
+    if (delim_pos == (int)(keys[0].length() - params.delim.length()))
+      keys[0].append("\xff");
+    else
+      keys[0].append(" ");
+  }
+
   rc = store->next_query_by_name(bucket_index_iname, keys, vals, params.prefix,
                                                                  params.delim);
   if (rc < 0) {
     ldpp_dout(dpp, 0) << "ERROR: NEXT query failed. " << rc << dendl;
     return rc;
   }
 
+  // Check if an extra key was fetched or not
+  if (rc == max) {
+    results.is_truncated = true;
+    results.next_marker = keys[max - 2];
```

Comment:
Did you verify that it does not cause the duplicates in the result?

Reply:
yes @andriytk, pls refer to the output section of issues I put in the PR description.

Reply:
I mean, when the number of objects is more than max.

Reply:
sorry I didn't get you probably..
In the above example, the no of objects is 3, and the max-keys requested are 2.

Reply:
`(rc == max)` where the `max was incremented by 1` than the user's request. 
This means we were successful in retrieving one extra key than someone requested. 
but in the response, we are not showing an extra key.
and the marker key is last key from the response, meaning 2nd last key from our vector.

Reply:
I see from the aws response, NextMarker is present in the response only if --delimiter is provided in the request.

```
#aws s3api list-objects --bucket sjain.bucket --profile aws --max-key 1 --prefix build --marker build1
{
    "IsTruncated": true,
    "Marker": "build1",
    "Contents": [
        {
            "Key": "build2",
            "LastModified": "2022-03-04T09:20:57.000Z",
            "ETag": "\"ec1f15371202fee4eda32d37b0c3415d\"",
            "Size": 13725,
            "StorageClass": "STANDARD",
            "Owner": {
                "ID": "6d0e3ec70bf0e6d90fe315b2becf38e295254485ff7f316a439dcb650ad5f117"
            }
        }
    ],
    "Name": "sjain.bucket",
    "Prefix": "build",
    "MaxKeys": 1,
    "EncodingType": "url"
}
```

with --delimiter in the request - 
```
#aws s3api list-objects --bucket sjain.bucket --profile aws --max-key 1 --prefix build --marker build1 --delimiter l
{
    "IsTruncated": true,
    "Marker": "build1",
    "NextMarker": "build2",
    "Contents": [
        {
            "Key": "build2",
            "LastModified": "2022-03-04T09:20:57.000Z",
            "ETag": "\"ec1f15371202fee4eda32d37b0c3415d\"",
            "Size": 13725,
            "StorageClass": "STANDARD",
            "Owner": {
                "ID": "6d0e3ec70bf0e6d90fe315b2becf38e295254485ff7f316a439dcb650ad5f117"
            }
        }
    ],
    "Name": "sjain.bucket",
    "Prefix": "build",
    "Delimiter": "l",
    "MaxKeys": 1,
    "EncodingType": "url"
}
```

Is this the case with our logic? I guess no. 
Does it make sense not to provide NextMarker in the response if --delimiter is not present in the request?

Reply:
no @sjain09 , there was only one key `build2` left after the marker `build1` with prefix `build`.
am I missing something?

Reply:
Looks like the marker is the key from which start to fetch, but without including it in the result, right? It makes sense then (why `max - 2`).

Reply:
yes correct, the NextMarker will be the last key from current response & 
when we will mention that key as a Marker to the next query the key will be omitted from the results.
max-2 as NextMarker because, we have fetched an additional key at max-1 by incrementing the max itself, just to know if is_truncated evaluates to true or false. We don't want to show that key to user. that key will be shown in the next query.